File: interpreter_first/src/com/craftinginterpreters/lox/Scanner.java
package com.craftinginterpreters.lox;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import static com.craftinginterpreters.lox.TokenType.*;

class Scanner {
    private final String source;
    private final List<Token> tokens = new ArrayList<>();
    private int start = 0; // first character in the lexeme being scanned
    private int current = 0; // the character in the lexeme being scanned
    private int line = 1; // tracks source line

    Scanner(String source) {
        this.source = source;
    }

    List<Token> scanTokens() {
        while (!isAtEnd()) {
            start = current;
            scanToken();
        }

        tokens.add(new Token(EOF, "", null, line));
        return tokens;
    }

    private boolean isAtEnd() {
        return current >= source.length();
    }
}

File: c/scanner.c
#include <stdio.h>
#include <string.h>

#include "common.h"
#include "scanner.h"

//
/*
 * Scanner chews through the users source code, it tracks how far it's gone.
 * start: pointer marks the beginning of the current lexeme being scanned.
 * current: pointer marks the character currently being looked at.
 * */
typedef struct {
    const char* start;
    const char* current;
    int line;
} Scanner;


Scanner scanner;

void initScanner(const char* source) {
    scanner.start = source;
    scanner.current = source;
    scanner.line = 1;
}